### DISCUSSION TITLE
[Gathering] Efficient Rendering / Optimizing

### DISCUSSION THREAD

**Discussion by DasLixou** [vizia/vizia#393] [Development]
This is just an issue to together gather all requirements and ideas to optimize on rendering. I'll start with sharing my core thought.

# Optimizing Rendering

So what we want is basicly not just to re-render everything when something changes. And I'll demonstate my though with the spotify desktop app for example usage. Basicly, when you play a song and are in the playlist, there are two things which will cause a re-render: An animated icon next to the current song and a the progress bar with the current time (ok maybe count that as 3 views lol). So when you think about writing this in vizia and want to optimize it, the first thing we need vizia to tell is that our component has changed and need a rerender. Currently this looks like this from the user perspective:
```rust
cx.needs_redraw();
```
In code this works like this:
```rust
// src/context/event.rs:339
pub fn needs_redraw(&mut self) {
    self.style.needs_redraw();
}
// you might think that the style is linked to the element, but its a global style
// src/style/mod.rs:1144
pub fn needs_redraw(&mut self) {
    self.system_flags.set(SystemFlags::REDRAW, true);
}
```
So in short: currently, we only know that a we need to redraw, we DON'T KNOW WHAT to redraw.
For this, I see two basic things:
1. passing an `Entity` as id to the needs_redraw method:
 ```rust
cx.needs_redraw(cx.current());
```
2. As you might see and think: "Wait, but won't we just always pass `cx.current()` in there?". So we could make the standard take itself and then make a special second function for this:
 ```rust
cx.needs_redraw(); // it now stores it together with cx.current()
// or
cx.needs_redraw_extern(other_entity);
```
> Any thoughts and other suggestions welcome :D

For detailed implementation about rendering, I'll from now on split this issue in 2 subgenres: Redraw requested by just one entity/view and redraw request from multiple entities/views.

## Optimizing on redraw request by just one Entity/View

This should be pretty simple. I am not sure how other UI Frameworks solve this, but I got a pretty simple and beautiful solution i would say. So lets get back to the spotify example and for now just focus on the animated icon. When you have some knowledge in rendering, you know the principle of clearing. In short: When we would only redraw the animated icon, it would overlap on the old state and at some time you wouldnt see it animate because you see all old states under it. So for that, we would have to render everything that is under the view to render first, so we get a result where everything is there EXCEPT the thing we want to render. Perfect! Now we can just simply render our animation with the new state without seeing the old one... But... there's a problem. See the background? If we would just straight render it again, every song that was above the background now is below it and we would need to rerender it. Stupid, right? Well, there's a neat thing in femtovg, the render framework behind vizia, which is called a `Scissor`. With that, we can limit the rendering to a specific section. So we'll rerender everything below our animated icon with a scissor cropped only to the boundings of our view that will be renrendered, then render it. Now it's perfect! So lets summarize this in simple steps:

### Summary

1. Get the bounding box of the View which should be redrawn.
2. Instead of redrawing element, use the old iterator but sort them out by checking if they intersect with the bounding box of 1.
3. Rendering with the Scissor of bounding box in 1.

Now we can go a little bit further...

## Optimizing on redraw request from multiple entities/views.

Your first thought might be: just use the same technique as above. And in some point you are right. But there's a slight issue: A scissor is only a rectangle. So when we would want to redraw 2 views which both are on top of a background which is like.. always... then we need to iterate over all entites twice, render them twice, one time with scissor from the first element, then with the scissor from the second element. So we iterate and render TWICE. That is everything but optimized. So for this I only see one possibility: PR. Contribute to femtovg to make scissors more customizable with weird shapes which are also not connected, because we need two rectangles as scissors which aren't like overlapping or so.
> I think such feature doesn't exist but correct me if I'm wrong.

When this is not possible, then I would for now stick to redrawing everything when more than one view requested a redraw.

### Summary

1. Get the bounding boxes of the Views which should be redrawn.
2. Use the old iterator and sort the views out by checking if they intersect with one of the the bounding boxes of 1.
3. Rendering with a custom Scissor of bounding boxes in 1.

# Footnote
First of all, we also need to look what is with children of a view. When requesting a redraw just once you might think it would fall into the first mentioned case in this issue, but when it has children it is probably the second case, right? (im not sure either lol)

I'm happy to get feedback, thoughts on this idea and hopefully make vizia faster and better 😀. I know that its a little bit early, but we can still do a idea gathering, right?

And when its time to implement, I would be happy to help contribute 😁

**Comment by DasLixou** [comment]
Other things to consider:
* Layered Rendering (from discord conversation with @geom3trik )
* Sectioned(idk a name for that?) Requests => For animations do we really want to call redraw request every time ourselves? We could say it is animated and optimize with that information

**Comment by geom3trik** [comment]
Repeating here for transparency:

Ah cool 🙂 what you're describing is, more or less, what I had planned. It's something called dirty regions (it has other names too). Essentially you figure out which views need redrawing and then figure out a bounding rectangle which encompasses all of them, then redraw those views using scissor.

As you pointed out, redrawing a view can cause other views behind it to redraw as well, potentially multiple times (but this can still be worth it overall). There's a number of tradeoffs and decisions to be made here. E.g. when is it more optimal to just rerender everything. Other similar ideas to consider are using tiles instead of dirty regions (a view may render in more than one tile), or a layered approach. The difficulty with layers is that a render target switch (rendering to an image and then back to the screen) can be quite expensive.
